src/managers/GoogleCalendarManager.ts

```diff
@@ -0,0 +1,196 @@
+
+/**
+ * GoogleCalendarManager manages automatic event notifications on Discord of events on
+ * the calendar that are occuring 1 hour from the current time or at the current time.
+ */
+@Service()
+export default class {
+  /**
+   * The Google Calendar Client that allows us to interact with the API.
+   */
+  private calendar: calendar_v3.Calendar;
+
+  /**
+   * Cronjob to check for upcoming meetings every 15 minutes.
+   */
+  public meetingNotificationsJob!: schedule.Job;
+
+  /**
+   * List of all calendarIDs to search through when checking for upcoming events.
+   */
+  public calendarList: string[];
+
+  /**
+   * Mapping from each calendarID to the specific channel/people to notify for each event.
+   */
+  public calendarMapping: Map<string, DiscordInfo>;
+
+  /**
+   * Updates the Google Calendar auth and client before an API call.
+   * @param client The original client, for access to the configuration.
+   */
+  private async refreshAuth(client: BotClient) : Promise<void> {
+    const auth = new google.auth.GoogleAuth({
+      keyFilename: client.settings.googleSheetsKeyFile, 
+      scopes: ['https://www.googleapis.com/auth/calendar.readonly', 'https://www.googleapis.com/auth/calendar'],
+    });
+    const authClient = await auth.getClient();
+    this.calendar = google.calendar({ version: 'v3', auth: authClient });
+  }
+
+  /**
+   * Queries all calendars for events happening in the given time window and sends
+   * messages to the calendar's respective Discord channel for each meeting.
+   * @param client The original client, for access to the configuration.
+   * @param start Date object storing the start of the time window we are querying for.
+   */
+  private async sendMeetingPings(client: BotClient, start: DateTime): Promise<void> {
+    /**
+     * The end of the time window of the query is one minute after the given start time.
+     */
+    const end = start.plus({ minutes: 1 });
+    // Reading in calendar colors from the Google API to color our embeds
+    const colors = await this.calendar.colors.get({});
+    /**
+     * Checking through all calendars in our calendar list...
+     */
+    this.calendarList.map(async calendarID => {
+      const res = await this.calendar.events.list({
+        calendarId: calendarID,
+        timeMin: start.toISO(),
+        timeMax: end.toISO(),
+        singleEvents: true,
+        orderBy: 'startTime',
+      });
+  
+      // Check that there are events to send notifications for
+      if (res && res.data.items) {
+        const events = res.data.items;
+        // Getting the specific color ID of our current calendar to color our embeds later.
+        const calColorResponse = await this.calendar.calendarList.get({ calendarId: calendarID });
+        const calColorID = calColorResponse.data.colorId;
+        events.map(async (event, i) => {
+          if (event && event.start && event.end && event.start.dateTime && event.end.dateTime) {
```

Comment:
Is there an instance where an event returned by the Google Calendar query would not have these fields defined?

I figure this is a type guard so that TypeScript doesn't complain about these fields being undefined, but I want to make sure that is the case.

Reply:
All events returned by the Google Calendar are guaranteed to have start and end info attached to them if the event is in the event list, so I put the line there just to stop TypeScript from complaining about them being potentially undefined.